packages/compass-components/src/components/document-list/element.tsx

```diff
@@ -0,0 +1,508 @@
+      valid: isValid,
+      validationMessage: !isValid ? el.invalidTypeMessage ?? null : null,
+    },
+    type: {
+      value: el.currentType,
+      change(newVal: HadronElementType['type']) {
+        el.changeType(newVal);
+      },
+    },
+    revert: el.isRevertable() ? el.revert.bind(el) : null,
+    remove: el.isNotActionable() ? null : el.remove.bind(el),
+    expandable: Boolean(el.elements),
+    children: el.elements ? [...el.elements] : [],
+    level: el.level,
+    parentType: el.parent?.currentType,
+    removed: el.isRemoved(),
+  };
+}
+const buttonReset = css({
+  margin: 0,
+  padding: 0,
+  border: 'none',
+  background: 'none',
+});
+
+const hadronElement = css({
+  display: 'flex',
+  paddingLeft: spacing[2],
+  paddingRight: spacing[2],
+  '&:hover': {
+    backgroundColor: uiColors.gray.light2,
+  },
+});
+
+const elementInvalid = css({
+  backgroundColor: uiColors.yellow.light3,
+  '&:hover': {
+    backgroundColor: uiColors.yellow.light2,
+  },
+});
+
+const elementRemoved = css({
+  backgroundColor: uiColors.red.light3,
+  '&:hover': {
+    backgroundColor: uiColors.red.light2,
+  },
+});
+
+const elementActions = css({
+  flex: 'none',
+  width: spacing[3],
+});
+
+const elementLineNumber = css({
+  flex: 'none',
+  position: 'relative',
+  marginLeft: spacing[1],
+  boxSizing: 'content-box',
+});
+
+const addFieldActionsContainer = css({
+  position: 'absolute',
+  top: 0,
+  right: 0,
+});
+
+const lineNumberCount = css({
+  '&::before': {
+    display: 'block',
+    width: '100%',
+    counterIncrement: 'line-number',
+    content: 'counter(line-number)',
+    textAlign: 'end',
+    color: uiColors.gray.base,
+  },
+});
+
+const lineNumberInvalid = css({
+  backgroundColor: uiColors.yellow.base,
+  '&::before': {
+    color: uiColors.yellow.dark2,
+  },
+});
+
+const lineNumberRemoved = css({
+  backgroundColor: uiColors.red.base,
+  color: uiColors.red.light3,
+  '&::before': {
+    color: uiColors.red.light3,
+  },
+});
+
+const elementSpacer = css({
+  flex: 'none',
+});
+
+const elementExpand = css({
+  width: spacing[3],
+  flex: 'none',
+});
+
+const elementKey = css({
+  flex: 'none',
+  fontWeight: 'bold',
+  maxWidth: '70%',
+});
+
+const elementDivider = css({
+  flex: 'none',
+  userSelect: 'none',
+});
+
+const elementValue = css({
+  flex: 1,
+  minWidth: 0,
+  maxWidth: '100%',
+});
+
+const elementType = css({
+  flex: 'none',
+  marginLeft: spacing[1],
+});
+
+const actions = css({
+  display: 'none',
+});
+
+const actionsVisible = css({
+  // We are deliberately not using useFocus and useHover hooks in the component
+  // as these listeners are expensive and so with the amount of code and DOM on
+  // the screen causes noticeable frame skips in the browser. This code is a bit
+  // brittle as we can't really reference the exact className of an emotion
+  // style so we have to rely on data attributes, but the chances that this
+  // should ever be an issue are pretty slim
```

Comment:
This should address the initial input lag. A bit hacky to reference parent element like that, but works well and improves performance in this hotspot a lot